src/util/forall.hpp

```diff
@@ -80,6 +80,8 @@ void forall(gpu, int begin, int end, LOOP_BODY&& body)
   hipLaunchKernelGGL(forall_kernel_gpu, dim3(gridSize), dim3(blockSize), 0,0,
                      begin, end - begin, body);
   hipDeviceSynchronize();
+#elif defined(CHAI_ENABLE_GPU_SIMULATION_MODE)
```

Comment:
This check should be first. If you have cuda enabled and the gpu simulation mode enabled, the simulation mode check will never get reached.